File: Algorithms/Source Code Algorithms/Sorting/HeapSort/SCHeapSort.java
void sort(int[] a) {
    for (int i = a.length / 2 - 1; i >= 0; i--)
        heapify(a, a.length, i);
    for (int i = a.length - 1; i >= 0; i--) {
        int tempA0 = a[0];
        a[0] = a[i];
        a[i] = tempA0;
        heapify(a, i, 0);
    }
}

void heapify(int[] a, int n, int i) {
    int largest = i;
    int l = 2 * i + 1;
    int r = 2 * i + 2;
    if (l < n && a[l] > a[largest])
        largest = l;
    if (r < n && a[r] > a[largest])
        largest = r;
    if (largest != i) {
        int tempAI = a[i];
        a[i] = a[largest];
        a[largest] = tempAI;
        heapify(a, n, largest);
    }
}

File: Algorithms/Source Code Algorithms/Sorting/HeapSort/SCHeapSort.swift
func sort(_ a: inout [Int]) {
    var i: Int = a.count / 2 - 1
    while i >= 0 {
        heapify(a: &a, n: a.count, i: i)
        i -= 1
    }
    i = a.count - 1
    while i >= 0 {
        let tempA0 = a[0]
        a[0] = a[i]
        a[i] = tempA0
        heapify(a: &a, n: i, i: 0)
        i -= 1
    }
}

func heapify(a: inout [Int], n: Int, i: Int) {
    var largest = i
    let l = 2 * i + 1
    let r = 2 * i + 2
    if l < n && a[l] > a[largest] {
        largest = l
    }
    if r < n && a[r] > a[largest] {
        largest = r
    }
    if largest != i {
        let tempAI = a[i]
        a[i] = a[largest]
        a[largest] = tempAI
        heapify(a: &a, n: n, i: largest)
    }
}
